linux/src/threadpool_linux.c

```diff
@@ -0,0 +1,455 @@
+        srw_lock_destroy(threadpool->srw_lock);
+        free(threadpool);
+    }
+}
+
+int threadpool_schedule_work(THREADPOOL_HANDLE threadpool, THREADPOOL_WORK_FUNCTION work_function, void* work_function_context)
+{
+    int result;
+    if (threadpool == NULL || work_function == NULL)
+    {
+        LogError("Invalid arguments: THREADPOOL_HANDLE threadpool: %p, THREADPOOL_TASK_FUNC work_function: %p", threadpool, work_function);
+        result = MU_FAILURE;
+    }
+    else
+    {
+        if (interlocked_add(&threadpool->state, 0) != (int32_t)THREADPOOL_STATE_OPEN)
+        {
+            LogWarning("Threadpool schedule work called on non-open pool");
+            result = MU_FAILURE;
+        }
+        else
+        {
+            //int64_t insert_pos = (interlocked_increment_64(&threadpool->insert_idx) % threadpool->task_array_count) - 1;
+            int64_t insert_pos = interlocked_exchange_64(&threadpool->insert_idx, threadpool->insert_idx+1) % threadpool->task_array_count ;
+
+            // Do we need to reallocate or do we need to roll over
+            int32_t task_state = interlocked_add(&threadpool->task_array[insert_pos].task_state, 0);
+            if (task_state != TASK_NOT_USED && task_state != TASK_RESULT_INVALID)
+            {
+                // The first task is not complete, so we just need to reallocate and
+                // go to the next item
+                if (reallocate_threadpool_array(threadpool) != 0)
+                {
+                    LogError("Failure reallocating threadpool");
+                    result = MU_FAILURE;
+                }
+                else
+                {
+                    insert_pos = interlocked_add_64(&threadpool->insert_idx, 0) - 1;
+                    result = 0;
+                }
+            }
+            else
+            {
+                result = 0;
+            }
+
+            if (result == 0)
+            {
+                srw_lock_acquire_shared(threadpool->srw_lock);
+                THREADPOOL_TASK* task_item = &threadpool->task_array[insert_pos];
+                task_item->task_param = work_function_context;
+                task_item->task_func = work_function;
+                (void)interlocked_exchange(&task_item->task_state, TASK_WAITING);
+                srw_lock_release_shared(threadpool->srw_lock);
+
+                do
+                {
+                    // Check to see if we should create another thread
+                    int32_t current_count = interlocked_add(&threadpool->thread_count, 0);
+                    if (current_count < threadpool->max_thread_count)
+                    {
+                        if (interlocked_compare_exchange(&threadpool->thread_count, current_count+1, current_count) == current_count)
+                        {
+                            if (ThreadAPI_Create(&threadpool->thread_handle_list[threadpool->list_index], threadpool_work_func, threadpool) != THREADAPI_OK)
```

Comment:
I think these need to be reversed. But maybe we should make it a moot point and remove the dynamic thread allocation for now.